app/Foliage/PrepareSource.hs

```diff
@@ -120,3 +88,21 @@ addPrepareSourceRule inputDir cacheDir = addBuiltinRule noLint noIdentity run
             liftIO $ rewritePackageVersion cabalFilePath pkgVersion
 
           return $ RunResult ChangedRecomputeDiff BS.empty srcDir
+
+    extractFromTarball tarballPath mSubdir outDir = do
+      withTempDir $ \tmpDir -> do
+        cmd_ "tar xzf" [tarballPath] "-C" [tmpDir]
+
+        ls <-
+          -- remove "." and ".."
+          filter (not . all (== '.'))
+            -- NOTE: Don't let shake look into tmpDir! it will cause
+            -- unnecessary rework because tmpDir is always new
+            <$> liftIO (IO.getDirectoryContents tmpDir)
+
+        -- Special treatment of top-level directory: we remove it
+        let byPassSingleTopLevelDir = case ls of [l] -> (</> l); _ -> id
+            applyMSubdir = case mSubdir of Just s -> (</> s); _ -> id
+            targetDir = applyMSubdir $ byPassSingleTopLevelDir tmpDir
+
+        cmd_ "cp --recursive --no-target-directory --dereference" [targetDir, outDir]
```

Comment:
IDEM for `cpInvocation`?

Reply:
E.g., we can document https://github.com/input-output-hk/foliage/issues/47#issuecomment-1489530344